k8s/cmd/nsmdp/nsmdp.go

```diff
@@ -231,6 +226,78 @@ func NewNSMDeviceServer(serviceRegistry serviceregistry.ServiceRegistry) error {
 	return err
 }
 
+func (nsm *nsmClientEndpoints) addClientDevice() {
+	nsm.mutext.Lock()
+	defer nsm.mutext.Unlock()
+
+	for {
+		nsm.clientId++
+
+		id := fmt.Sprintf("nsm-%d", nsm.clientId)
+		if _, ok := nsm.devs[id]; ok {
+			// Item already exists, increment to new one.
+			continue
+		}
+		// Add a new device Id
+		dev := &pluginapi.Device{
+			ID:     id,
+			Health: pluginapi.Healthy,
+		}
+		nsm.devs[id] = dev
+		nsm.resp.Devices = append(nsm.resp.Devices, dev)
+		break
+	}
+
+}
+
+func (n *nsmClientEndpoints) sendDeviceUpdate() {
+	n.mutext.Lock()
+	defer n.mutext.Unlock()
+	if n.pluginApi != nil {
+		api := *n.pluginApi
+		logrus.Infof("Send device plugins intfo %v", n.resp)
+		if err := api.Send(n.resp); err != nil {
+			logrus.Errorf("Failed to send response to kubelet: %v\n", err)
+		}
+	}
+}
+
+func (n *nsmClientEndpoints) reciveWorkspaces() {
+	for {
+		reply, err := enumWorkspaces(n.serviceRegistry)
+		if err != nil {
+			logrus.Errorf("Error receive devices from NSM. %v", err)
+			// Make a fast delay to faster startup of NSMD.
+			<-time.Tick(100 * time.Millisecond)
+			continue
+		}
```

Comment:
you actually can remove the `continue` and move the end of the for cycle with `break` here. The rest of the code is not repeated in any case.

Reply:
Nope, If we remove continue here, it will be delay 30seconds before kubelet will receive a initial list of devices, since all of previous ones could be already used by nsmd by restore. This will slowdown all test in case NSMD will not yet start. So this continue allow to wait for NSMD startup, like 3-5 seconds before we will do a huge delay.